presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +80,261 @@
+        this.jettyIoPool = new JettyIoPool("presto-jdbc", new JettyIoPoolConfig());
+        this.cacheLock = new ReentrantReadWriteLock();
+        this.queryExecutorCleanupService = new ScheduledThreadPoolExecutor(1);
+
+        this.queryExecutorCache = CacheBuilder.newBuilder()
+                .removalListener(new RemovalListener<Map<Object, Object>, QueryExecutorHolder>() {
+                        @Override
+                        public void onRemoval(RemovalNotification<Map<Object, Object>, QueryExecutorHolder> notification)
+                        {
+                            notification.getValue().closeQueryExecutor();
+                        }
+                    })
+                .build(new CacheLoader<Map<Object, Object>, QueryExecutorHolder>() {
+                        @Override
+                        public QueryExecutorHolder load(Map<Object, Object> clientProperties)
+                        {
+                            return new QueryExecutorHolder(
+                                    QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION, jettyIoPool),
+                                    clientProperties);
+                        }
+                    });
     }
 
     @Override
     public void close()
     {
-        queryExecutor.close();
+        if (closed.compareAndSet(false, true)) {
+            queryExecutorCleanupService.shutdown();
+            queryExecutorCache.invalidateAll();
+            jettyIoPool.close();
+        }
+    }
+
+    private class QueryExecutorHolder
+    {
+        private QueryExecutor queryExecutor;
+
+        private Map<Object, Object> clientProperties;
+
+        private AtomicInteger refcount = new AtomicInteger(0);
+
+        private QueryExecutorHolder(QueryExecutor queryExecutor, Map<Object, Object> clientProperties)
+        {
+            this.queryExecutor = requireNonNull(queryExecutor, "wrapped is null");
+            this.clientProperties = requireNonNull(clientProperties, "clientProperties is null");
+        }
+
+        private void closeQueryExecutor()
+        {
+            queryExecutor.close();
+        }
+
+        public void release()
+        {
+            Runnable deferredClose = () -> {
+                cacheLock.writeLock().lock();
+                try {
+                    if (refcount.decrementAndGet() == 0) {
+                        queryExecutorCache.invalidate(clientProperties);
+                    }
+                }
+                finally {
+                    cacheLock.writeLock().unlock();
+                }
+            };
+
+            /*
+             * Defer cleanup into the future to avoid constantly creating and
+             * destroying QE's in the case where a single thread is repeatedly
+             * calling connect() and close() for the same server.
+             */
+            queryExecutorCleanupService.schedule(deferredClose, 5, TimeUnit.MINUTES);
```

Comment:
For the workaround when clients fetches the connections and releases them very offen we can run out of memory because of storing the entire list of a `release` tasks for the last 5 minutes. Although those tasks don't take too much memory it is still better to cancel the previous tasks when the connection is aquired/released. Also you can schedule the release task only if the reference count is `0`.